### ISSUE TITLE
Warlock fails when many concurrent requests happen

### ISSUE THREAD

**Issue by ORESoftware** [TheDeveloper/warlock#22]
I am writing a library called` live-mutex`, it is much slower than using your library + Redis, and has a different use case, but apparently is more robust. Surely one of the biggest requirements of a locking library is to handle many concurrent requests to acquire a lock.

I have used this library in the past and it seemed to work, but since I had a test written for my library I decided to compare performance with other similar libraries.

Warlock should fail here, as is. If you change `each` to `eachSeries` it should succeed again.

You will need to start redis on default host/port and then run this:

https://github.com/ORESoftware/live-mutex/blob/master/test/speed/redis-warlock.js

Am I using the library incorrectly, or does Warlock fail upon many concurrent requests? Maybe there is a way to tune Warlock so it works in this case?

**Comment by calibr** [comment]
For me it once prints `unlocking...0` and many times `error => Could not acquire lock.`, it is ok because you are trying to acquire the same lock concurently before it is unlocked. Warlock doesn't wait until the lock is free, it triggers callback immediately after failed lock attempt.

**Comment by ORESoftware** [comment]
@calibr yeah, thanks for running the code. TBH it's not really a good design, and Warlock is not the first library that I have seen that does this. NPM lockfile library has the same exact problem https://github.com/npm/lockfile/issues/27.

if you only call lock() "in series", then it will work, but then what is the point of a locking library? lock requests are supposed to come in concurrently. Otherwise, you wouldn't even need to lock on something.

I guess you can handle the retries yourself in your own code - but what that means is you have to use polling your code! polling is not a good thing. On the other hand, Live-Mutex doesn't do this, it just fires a callback once the lock is available for a particular lock request.

TBH I am super confused as to why this is acceptable behavior for any locking library. There should always exist the option to wait until the lock is available for your particular request. If anything, you could pass a flag that determines whether you want to wait to acquire the lock or immediately callback if the lock could not be acquired.

**Comment by ORESoftware** [comment]
So with the text file contents that I just shared, I get this:

oleg@xps:~/foo$ node lib/blank.js /home/oleg/dogs.txt \\S+
dogs
oleg@xps:~/foo$ node lib/blank.js /home/oleg/dogs.txt \\S+
cats
oleg@xps:~/foo$ node lib/blank.js /home/oleg/dogs.txt \\S+
birds
oleg@xps:~/foo$ node lib/blank.js /home/oleg/dogs.txt \\S+
plants
oleg@xps:~/foo$ node lib/blank.js /home/oleg/dogs.txt \\S+
insects
oleg@xps:~/foo$ node lib/blank.js /home/oleg/dogs.txt \\S+
Error: EBADF: bad file descriptor, write

**Comment by calibr** [comment]
there is a method in warlock called `optimistic`, look at the usage example https://github.com/TheDeveloper/warlock#usage, it tries to acquire a lock through multiple attempts, it uses polling internally as well. Surely, push approach to wait for a lock is faster than poll, so I think you might want to compare your locking implementation with solutions that also use push, [zookeeper locks](https://zookeeper.apache.org/doc/r3.1.2/recipes.html#sc_recipes_Locks) for example.

**Comment by ORESoftware** [comment]
@calibr thanks for the tip, I didn't know about the optimistic option. Will check out zookeeper locks.

**Comment by ORESoftware** [comment]
@calibr I tried warlock.optimistic

TBH the results were pretty horrific, maybe you could fine-tune it to be less horrific but I doubt it will ever approach the performance of Live-Mutex

I just wrote this up yesterday
https://medium.com/@the1mills/a-better-mutex-for-node-js-4b4897fd9f11

Live-Mutex is pretty sweet when it comes down to it. It actually behaves like a locking library should.

Zookeeper was too heavy-weight for my purposes. I need a local solution, as explained my article - I needed a networked locking mechanism that I can use for a library - I can't expect the users of my library to install Redis or Zookeeper.

**Comment by germansokolov13** [comment]
So the issue is solved? It's still open. "Mutex not working with concurrent requests" is a rather sad prospect
